README.markdown
Name
====

lua-resty-requests - Yet Another HTTP Library for OpenResty.

![Build Status](https://travis-ci.org/tokers/lua-resty-requests.svg?branch=master)

```bash
resty -e 'print(require "resty.requests".get{ url = "https://github.com", stream = false }.content)'
```

Table of Contents
=================

* [Name](#name)
* [Status](#status)
* [Features](#features)
* [Synopsis](#synopsis)
* [Installation](#installation)
* [Methods](#methods)
    * [request](#request)
    * [state](#state)
    * [get](#get)
    * [head](#head)
    * [post](#post)
    * [put](#put)
    * [delete](#delete)
    * [options](#options)
    * [patch](#patch)
* [Response Object](#response-object)
* [Session](#session)
* [TODO](#todo)
* [Author](#author)
* [Copyright and License](#copyright-and-license)
* [See Also](#see-also)

Status
======

This Lua module now can be considered as production ready.

Note since the `v0.7.1` release, this module started using lua-resty-socket,
for working in the non-yieldable phases, but still more efforts are needed,
so **DONOT** use it in the `init` or `init_worker` phases (or other
non-yieldable phases).

Features
========

* HTTP/1.0, HTTP/1.1 and HTTP/2 (WIP).
* SSL/TLS support.
* Chunked data support.
* Convenient interfaces to support features like json, authorization and etc.
* Stream interfaces to read body.
* HTTP/HTTPS proxy.
* Latency metrics.
* Session support.

Synopsis
========

```lua
local requests = require "resty.requests"

-- example url
local url = "http://example.com/index.html"

local r, err = requests.get(url)
if not r then
    ngx.log(ngx.ERR, err)
    return
end

-- read all body
local body = r:body()
ngx.print(body)

-- or you can iterate the response body
-- while true do
--     local chunk, err = r:iter_content(4096)
--     if not chunk then
--         ngx.log(ngx.ERR, err)
--         return
--     end
--
--     if chunk == "" then
--         break
--     end
--
--     ngx.print(chunk)
-- end

-- you can also use the non-stream mode
-- local opts = {
--     stream = false
-- }
--
-- local r, err = requests.get(url, opts)
-- if not r then
--     ngx.log(ngx.ERR, err)
-- end
--
-- ngx.print(r.content)

-- or you can use the shortcut way to make the code cleaner.
local r, err = requests.get { url = url, stream = false }
```

Installation
============

* [LuaRocks](https://luarocks.org):

```bash
$ luarocks install lua-resty-requests
```

* [OPM](https://github.com/openresty/opm):

```bash
$ opm get tokers/lua-resty-requests
```

* Manually:

Just tweeks the `lua_package_path` or the `LUA_PATH` environment variable, to add the installation path for this Lua module:

```
/path/to/lua-resty-requests/lib/resty/?.lua;
```

Methods
=======

### request

**syntax**: *local r, err = requests.request(method, url, opts?)*  
**syntax**: *local r, err = requests.request { method = method, url = url, ... }

This is the pivotal method in `lua-resty-requests`, it will return a [response object](#response-object) `r`. In the case of failure, `nil`, and a Lua string which describles the corresponding error will be given.

The first parameter `method`, is the HTTP method that you want to use(same as
HTTP's semantic), which takes a Lua string and the value can be:

* `GET`
* `HEAD`
* `POST`
* `PUT`
* `DELETE`
* `OPTIONS`
* `PATCH`

The second parameter `url`, just takes the literal meaning(i.e. Uniform Resource Location),
for instance, `http://foo.com/blah?a=b`, you can omit the scheme prefix and as the default scheme,
`http` will be selected.

The third param, an optional Lua table, which contains a number of  options:

* `headers` holds the custom request headers.

* `allow_redirects` specifies whether redirecting to the target url(specified by `Location` header) or not when the status code is `301`, `302`, `303`, `307` or `308`.

* `redirect_max_times` specifies the redirect limits, default is `10`.

* `body`, the request body, can be:
    * a Lua string, or
    * a Lua function, without parameter and returns a piece of data (string) or an empty Lua string to represent EOF, or
    * a Lua table, each key-value pair will be concatenated with the "&", and Content-Type header will `"application/x-www-form-urlencoded"`

* `error_filter`, holds a Lua function which takes two parameters, `state` and `err`.
 the parameter `err` describes the error and `state` is always one of these values(represents the current stage):
    * `requests.CONNECT`
    * `requests.HANDSHAKE`
    * `requests.SEND_HEADER`
    * `requests.SEND_BODY`
    * `requests.RECV_HEADER`
    * `requests.RECV_BODY`
    * `requests.CLOSE`

You can use the method [requests.state](#state) to get the textual meaning of these values.


* `timeouts`, an array-like table, `timeouts[1]`, `timeouts[2]` and `timeouts[3]` represents `connect timeout`, `send timeout` and `read timeout` respectively (in milliseconds).

* `http10` specify whether the `HTTP/1.0` should be used, default verion is `HTTP/1.1`.
* `http20` specify whether the `HTTP/2` should be used, default verion is `HTTP/1.1`.

Note this is still unstable, caution should be exercised. Also, there are some
limitations, see [lua-resty-http2](https://github.com/tokers/lua-resty-http2) for the details.

* `ssl` holds a Lua table, with three fields:
  * `verify`, controls whether to perform SSL verification
  * `server_name`, is used to specify the server name for the new TLS extension Server Name Indication (SNI)

* `proxies` specify proxy servers, the form is like

```lua
{
    http = { host = "127.0.0.1", port = 80 },
    https = { host = "192.168.1.3", port = 443 },
}
```

When using HTTPS proxy, a preceding CONNECT request will be sent to proxy server.

* `hooks`, also a Lua table, represents the hook system that you can use to
manipulate portions of the request process. Available hooks are:
  * `response`, will be triggered immediately after receiving the response headers

you can assign Lua functions to hooks, these functions accept the [response object](#response-object) as the unique param.

```lua
local hooks = {
    response = function(r)
        ngx.log(ngx.WARN, "during requests process")
    end
}
```

Considering the convenience, there are also some "short path" options:

* `auth`, to do the Basic HTTP Authorization, takes a Lua table contains `user` and `pass`, e.g. when `auth` is:

```lua
{
    user = "alex",
    pass = "123456"
}
```

Request header `Authorization` will be added, and the value is `Basic YWxleDoxMjM0NTY=`.

* `json`, takes a Lua table, it will be serialized by `cjson`, the serialized data will be sent as the request body, and it takes the priority when both `json` and `body` are specified.

* `cookie`, takes a Lua table, the key-value pairs will be organized according to the `Cookie` header's rule, e.g. `cookie` is:

```lua
{
    ["PHPSESSID"] = "298zf09hf012fh2",
    ["csrftoken"] = "u32t4o3tb3gg43"
}
```

The `Cookie` header will be `PHPSESSID=298zf09hf012fh2; csrftoken=u32t4o3tb3gg43 `.

* `stream`, takes a boolean value, specifies whether reading the body in the stream mode, and it will be true by default. 

### state

**syntax**: *local state_name = requests.state(state)*

The method is used for getting the textual meaning of these values:

 * `requests.CONNECT`
 * `requests.HANDSHAKE`
 * `requests.SEND_HEADER`
 * `requests.SEND_BODY`
 * `requests.RECV_HEADER`
 * `requests.RECV_BODY`
 * `requests.CLOSE`

a Lua string `"unknown"` will be returned if `state` isn't one of the above values.

### get

**syntax**: *local r, err = requests.get(url, opts?)*  
**syntax**: *local r, err = requests.get { url = url, ... }*

Sends a HTTP GET request. This is identical with

```lua
requests.request("GET", url, opts)
```

### head
**syntax**: *local r, err = requests.head(url, opts?)*   
**syntax**: *local r, err = requests.head { url = url, ... }*

Sends a HTTP HEAD request. This is identical with

```lua
requests.request("HEAD", url, opts)
```

### post
**syntax**: *local r, err = requests.post(url, opts?)*    
**syntax**: *local r, err = requests.post { url = url, ... }*

Sends a HTTP POST request. This is identical with

```lua
requests.request("POST", url, opts)
```

### put
**syntax**: *local r, err = requests.put(url, opts?)*    
**syntax**: *local r, err = requests.put { url = url, ... }*

Sends a HTTP PUT request. This is identical with

```lua
requests.request("PUT", url, opts)
```

### delete
**syntax**: *local r, err = requests.delete(url, opts?)*  
**syntax**: *local r, err = requests.delete { url = url, ... }*

Sends a HTTP DELETE request. This is identical with

```lua
requests.request("DELETE", url, opts)
```

### options
**syntax**: *local r, err = requests.options(url, opts?)*  
**syntax**: *local r, err = requests.options { url = url, ... }*

Sends a HTTP OPTIONS request. This is identical with

```lua
requests.request("OPTIONS", url, opts)
```

### patch
**syntax**: *local r, err = requests.patch(url, opts?)*  
**syntax**: *local r, err = requests.patch { url = url, ... }*

Sends a HTTP PATCH request. This is identical with

```lua
requests.request("PATCH", url, opts)
```

Response Object
===============

Methods like `requests.get` and others will return a response object `r`, which can be manipulated by the following methods and variables:

* `url`, the url passed from caller
* `method`, the request method, e.g. `POST`
* `status_line`, the raw status line(received from the remote)
* `status_code`, the HTTP status code
* `http_version`, the HTTP version of response, e.g. `HTTP/1.1`
* `headers`, a Lua table represents the HTTP response headers(case-insensitive)
* `close`, holds a Lua function, used to close(keepalive) the underlying TCP connection
* `drop`, is a Lua function, used for dropping the unread HTTP response body, will be invoked automatically when closing (if any unread data remains)
* `iter_content`, which is also a Lua function, emits a part of response body(decoded from chunked format) each time called. 

This function accepts an optional param `size` to specify the size of body that the caller wants, when absent, `iter_content` returns `8192` bytes when the response body is plain or returns a piece of chunked data if the resposne body is chunked.

In case of failure, `nil` and a Lua string described the error will be returned.

* `body`, also holds a Lua function that returns the whole response body.

In case of failure, `nil` and a Lua string described the error will be returned.

* `json`, holds a Lua function, serializes the body to a Lua table, note the `Content-Type` should be `application/json`. In case of failure, `nil` and an error string will be given.

* `content`, the response body, only valid in the non-stream mode.

* `elapsed`, a hash-like Lua table which represents the cost time (in seconds) for each stage.
  * `elapsed.connect`, cost time for the TCP 3-Way Handshake;
  * `elapsed.handshake`, cost time for the SSL/TLS handshake (if any);
  * `elapsed.send_header`, cost time for sending the HTTP request headers;
  * `elapsed.send_body`, cost time for sending the HTTP request body (if any);
  * `elapsed.read_header`, cost time for receiving the HTTP response headers;
  * `elapsed.ttfb`, the time to first byte.

Note When HTTP/2 protocol is applied, the `elapsed.send_body` (if any) will be same as `elapsed.send_header`.

Session
=======

A session persists some data across multiple requests, like cookies data, authorization data and etc.

This mechanism now is still experimental.

A simple example:

```lua
s = requests.session()
local r, err = s:get("https://www.example.com")
ngx.say(r:body())
```

A session object has same interfaces with `requests`, i.e. those http methods.

TODO
====

* other interesting features...


Author
======

Alex Zhang (张超) zchao1995@gmail.com, UPYUN Inc.

Copyright and License
=====================

The bundle itself is licensed under the 2-clause BSD license.

Copyright (c) 2017-2019, Alex Zhang.

This module is licensed under the terms of the BSD license.

Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:

Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.
Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation and/or other materials provided with the distribution.
THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

See Also
=========

* upyun-resty: https://github.com/upyun/upyun-resty
* httpipe: https://github.com/timebug/lua-resty-httpipe


CHANGELOG.markdown
Table of Contents
=================

* [v0.7.3](#v0.7.3)
* [v0.7.2](#v0.7.2)
* [v0.7.2](#v0.7.2)
* [v0.7.1](#v0.7.1)
* [v0.7](#v0.7)
* [v0.6](#v0.6)
* [v0.5](#v0.5)
* [v0.4](#v0.4)
* [v0.3](#v0.3)


v0.7.3
======

> Date: 2019.07.18

* bugfix: treat Expect header value case-insensitive.
* test: download Test::Nginx with the super user permission.
* feature: shortcut calling for requests APIs.

v0.7.2
======

> Date: 2019.03.08

* doc: added the annotations about the compatibilty for the blocking phases.
* change: updated the lua-resty-socket dependency.
* bugfix: fixed the bootless keep-alive feature, now it works when the
condition is conformed.

v0.7.1
======

The old version is not so semantic, now we tweaked it.

> Date: 2019.02.21

This version just contains some minor improvements.

* doc: fix timeout misrepresent, seconds => milliseconds. Thanks jetz for the
patch.
* doc: README fix typo ("Authorization") 7 hours. Thanks Chris Kuehl for the
patch.
* doc: fixed some documentary typos.
* feature: used lua-resty-socket when the phase is not yieldable.
* feature: used cosocket pool backlog if the ngx_lua is new enough.

v0.7
====

> Date: 2018.12.03

This version refactored the logic about headers indexing, also fixed the relevant bug.

* refactored the request/response headers table logic.

v0.6
====

> Date: 2018.11.15

This version added some new features like r:json(), HTTPS proxy and some bugfixs.

* feature: supported HTTPS proxy based on HTTP CONNECT method.
* feature: r:json(), now one can get a Lua table which deserializes the response body from calling this method.
* feature: added a new option "use_default_type" to control whether adds a default content-type request header when request body exists.
* bugfix: the ttfb metric always records the "time to first header".
* bugfix: add "charset; utf-8" check to json response object "content-type" header, thanks Happy Totem for the report and pull request.

v0.5
====

> Date: 2018.10.25

This version has minor modifications but with a compatibilty broken change.

* change: Content-Length header will not be deleted even when users are using the function request body fashion.
* improve: now we don't launch ssl handshake if a reused connection is using.
* bugfix: http2.lua cannot be copied to the correct openresty lualib dir.

v0.4
====

> Date: 2018.10.09

This version supplemented more test cases and introduced the following
features.

* feature: supported the test of Expect request header.
* feature: supported installation from LuaRocks.
* feature: intergrated with lua-resty-http2, can use the HTTP/2 in the plain connection, this is still experimental.

v0.3
====

> Date: 2018.06.18

The first release.




lib/resty/requests/http.lua
-- Copyright (C) Alex Zhang

local cjson = require "cjson.safe"
local util  = require "resty.requests.util"

local tcp_sock   = ngx.socket.tcp
local ngx_match  = ngx.re.match
local ngx_gmatch = ngx.re.gmatch

local tab_concat = table.concat
local tab_insert = table.insert
local str_format = string.format
local str_lower  = string.lower
local str_find   = string.find
local str_byte   = string.byte
local str_sub    = string.sub

local pairs    = pairs
local ipairs   = ipairs
local tonumber = tonumber
local tostring = tostring

local CRLF     = "\r\n"
local SEP_BYTE = ("/"):byte(1)

local HTTP_MOVED_PERMANENTLY = ngx.HTTP_MOVED_PERMANENTLY
local HTTP_MOVED_TEMPORARILY = ngx.HTTP_MOVED_TEMPORARILY
local HTTP_OK                = ngx.HTTP_OK
local HTTP_NOT_MODIFIED      = ngx.HTTP_NOT_MODIFIED
local HTTP_NO_CONTENT        = ngx.HTTP_NO_CONTENT


local _M = { _VERSION = "0.1" }

local STATE = {
    UNREADY     = -1,
    READY       = 0,
    CONNECT     = 1,
    HANDSHAKE   = 2,
    SEND_HEADER = 3,
    SEND_BODY   = 4,
    RECV_HEADER = 5,
    RECV_BODY   = 6,
    CLOSE       = 7,
}

local STATE_NAME = {
    [STATE.UNREADY] = "unready",
    [STATE.READY] = "ready",
    [STATE.CONNECT] = "connect",
    [STATE.HANDSHAKE] = "handshake",
    [STATE.SEND_HEADER] = "send_header",
    [STATE.SEND_BODY] = "send_body",
    [STATE.RECV_HEADER] = "recv_header",
    [STATE.RECV_BODY] = "recv_body",
    [STATE.CLOSE] = "close",
}

local HTTP_METHODS = {
    "GET",
    "HEAD",
    "POST",
    "PUT",
    "DELETE",
    "OPTIONS",
    "PATCH",
}

local DEFAULT_PORT = {
    http  = 80,
    https = 443,
}

local DEFAULT_SCHEME = "http"
local DEFAULT_PATH   = "/"


local function parse_url(url)
    if not util.is_str(url) then
        return
    end

    local m, err = ngx_match(url, [[(?:(https?)://)?([^:/]+)(?::(\d+))?(.*)]],
                             "jo")
    if not m then
        return nil, err
    end

    local parts = util.new_tab(0, 5)

    parts.scheme = m[1] or DEFAULT_SCHEME
    parts.host = m[2]
    parts.port = m[3] or DEFAULT_PORT[parts.scheme]

    if not m[4] or m[4] == "" then
        parts.path = DEFAULT_PATH
        parts.args = ""

    else
        local colon = str_find(m[4], "?")
        if colon then
            parts.path = str_sub(m[4], 1, colon - 1)
            parts.args = str_sub(m[4], colon + 1)
        else
            parts.path = m[4]
            parts.args = ""
        end

        if parts.path == "" then
            parts.path = DEFAULT_PATH
        end
    end

    return parts
end


local function adjust_location(r)
    local location = r.headers["Location"]
    if not util.is_str(location) then
        return
    end

    local prefix = str_byte(location, 1, 1)

    if prefix ~= SEP_BYTE then
        return location
    end

    -- relative
    local url_parts = r.ctx.url_parts
    return str_format("%s://%s:%d%s", url_parts["scheme"], url_parts["host"],
                      url_parts["port"], location)
end


local function serialize(headers)
    local t = util.new_tab(4, 0)
    for k, v in pairs(headers) do
        tab_insert(t, str_format("%s: %s", tostring(k), tostring(v)))
    end

    tab_insert(t, CRLF)

    return tab_concat(t, CRLF)
end


local function no_body(r)
    if not r then
        return true
    end

    local status_code = r.status_code

    -- 1xx, 204 and 304
    if status_code < HTTP_OK
       or status_code == HTTP_NO_CONTENT
       or status_code == HTTP_NOT_MODIFIED then
        return true
    end

    -- HEAD
    if r.method == "HEAD" then
        return true
    end
end


local function allow_redirects(r)
    local config = r.ctx.config
    if not config.allow_redirects then
        return
    end

    config.redirect_max_times = config.redirect_max_times - 1

    if config.redirect_max_times <= 0 then
        return
    end

    local status_code = r.status_code
    if status_code ~= HTTP_MOVED_PERMANENTLY
       and status_code ~= HTTP_MOVED_TEMPORARILY
    then
        return false
    end

    return true
end


local function parse_header_line(line)
    if not line or #line < 3 then
        return
    end

    local m, err = ngx_match(line, "^(.+?):\\s*(.+)$", "jo")
    if not m then
        return nil, nil, err
    end

    return m[1], m[2]
end


local function parse_status_line(status_line)
    local m, err = ngx_match(status_line, "HTTP/(.+?)\\s.*?(\\d+).*", "jo")
    if not m then
        return nil, err
    end

    return {
        status_code = tonumber(m[2]),
        http_version = m[1],
    }
end


local function adjust_request_headers(ctx)
    local config = ctx.config
    local host = ctx.url_parts.host
    local body = config.body
    local json = config.json
    local headers = config.headers

    headers["Content-Length"] = nil
    headers["Transfer-Encoding"] = nil

    -- lua table
    if json then
        body = cjson.encode(json)
        config.body = body

        if not config.headers["Content-Type"] then
            headers["Content-Type"] = "application/json"
        end
    end

    -- message length
    if util.is_func(body) then
        headers["Transfer-Encoding"] = "chunked"
        headers["Content-Type"] = "application/octet-stream"
    elseif body then
        headers["Content-Length"] = #body
        headers["Content-Type"] = "text/plain"
    end

    -- Host
    if not headers["Host"] then
        headers["Host"] = host
    end

    -- Connection
    headers["Connection"] = ctx.sessoin and "keep-alive" or "close"

    local auth = config.auth
    if auth then
        headers["Authorization"] = auth
    end

    local cookie = config.cookie
    if cookie then
        local plain = util.new_tab(4, 0)
        for k, v in pairs(cookie) do
            tab_insert(plain, ("%s=%s"):format(tostring(k), tostring(v)))
        end

        headers["Cookie"] = tab_concat(plain, "; ")
    end
end


local function adjust_response_headers(r)
    for k, v in pairs(r.headers) do
        if util.is_tab(v) then
            r.headers[k] = tab_concat(v, ",")
        end
    end

    local cookies = r.headers["Cookie"]
    if not cookies then
        return true
    end
end


local function adjust_raw(r)
    local eof
    local rest = tonumber(r.headers["Content-Length"])

    return function(r, size)
        local ctx = r.ctx
        if ctx.state == STATE.CLOSE then
            return nil, "closed"
        end

        if eof then
            return nil, "eof"
        end

        -- empty string will be yielded when no body
        if no_body(r) then
            eof = true
            return ""
        end

        ctx.state = STATE.RECV_BODY

        if rest == 0 then
            eof = true
            return ""
        end

        size = size or 8192

        if rest and rest < size then
            size = rest
        end

        local data, err = ctx.sock:receive(size)
        if not data then
            local error_filter = ctx.config.error_filter
            if error_filter then
                error_filter(STATE.RECV_BODY, err)
            end

            r:close()
            return nil, err
        end

        if rest then
            rest = rest - #data
        end

        return data
    end
end


local function adjust_chunked()
    local eof

    return function(r, size)
        local ctx = r.ctx
        if ctx.state == STATE.CLOSE then
            return nil, "closed"
        end

        if ctx.chunk.leave then
            eof = true
            ctx.chunk.leave = false
            return ""
        end

        if eof then
            return nil, "eof"
        end

        if no_body(r) then
            eof = true
            return ""
        end

        ctx.state = STATE.RECV_BODY

        local chunk = ctx.chunk

        local t
        if size then
            t = util.new_tab(0, 4)
        end

        while true do
            if chunk.rest == 0 then
                -- read next chunk
                local size, err = ctx.line_reader()
                if not size then
                    local error_filter = ctx.config.error_filter
                    if error_filter then
                        error_filter(STATE.RECV_BODY, err)
                    end

                    return nil, err
                end

                -- just ignore the chunk-extensions
                local ext = size:find(";")
                if ext then
                    size = size:sub(1, ext - 1)
                end

                size = tonumber(size, 16)
                if not size then
                    r:close()
                    return nil, "invalid chunk header"
                end

                chunk.size = size
                chunk.rest = size
            end

            -- end
            if chunk.size == 0 then
                if not size then
                    eof = true
                    return ""
                end

                chunk.leave = true

                break
            end

            local read_size = size
            if not size or read_size > chunk.rest then
                read_size = chunk.rest
            end

            local data, err = ctx.sock:receive(read_size)
            if err then
                local error_filter = ctx.config.error_filter
                if error_filter then
                    error_filter(STATE.RECV_BODY, err)
                end

                r:close()

                return data, err
            end

            if not size then
                chunk.rest = 0
            else
                size = size - read_size
                chunk.rest = chunk.rest - read_size
            end

            if chunk.rest == 0 then
                local dummy, err = ctx.line_reader()
                if dummy ~= "" then
                    if err then
                        local error_filter = ctx.config.error_filter
                        if error_filter then
                            error_filter(STATE.RECV_BODY, err)
                        end
                    end

                    return nil, err or "invalid chunked data"
                end
            end

            if not size then
                return data
            end

            tab_insert(t, data)
            if size == 0 then
                break
            end
        end

        return tab_concat(t, "")
    end
end


local function adjust_body(r)
    -- 8) iter_content
    local chunked = r.headers["Transfer-Encoding"]
    if not chunked or not str_find(chunked, "chunked") then
        r.iter_content = adjust_raw(r)
    else
        r.iter_content = adjust_chunked()
    end


    -- 9) body
    r.body = function(r)
        local t = util.new_tab(4, 0)
        while true do
            local data, err = r:iter_content()
            if err then
                return nil, err
            end

            if data == "" then
                break
            end

            tab_insert(t, data)
        end

        return tab_concat(t, "")
    end

    -- 10) json
    r.json = function(r)
        local data, err = r:body()
        if not data then
            return nil, err
        end

        return cjson.encode(data)
    end
end


local function close_sock(r)
    local ctx = r.ctx
    local config = ctx.config
    local sock = ctx.sock
    local error_filter = config.error_filter

    ctx.state = STATE.CLOSE
    ctx.sock = nil

    local connection = r.headers["Connection"]
    local ok, err

    -- the persistent connection is the default behaviour in HTTP/1.1
    if not connection and r.http_version == "1.1" then
        ok, err = sock:setkeepalive()

    elseif str_lower(connection) == "keep-alive" then
        ok, err = sock:setkeepalive()

    else
        ok, err = sock:close()
    end

    if error_filter and not ok then
        error_filter(ctx.state, err)
    end
end


local function connect(ctx)
    ctx.state = STATE.CONNECT

    local sock = tcp_sock()
    local config = ctx.config
    local url_parts = ctx.url_parts

    sock:settimeout(config.timeouts[1])

    local host   = url_parts.host
    local port   = url_parts.port
    local scheme = url_parts.scheme

    -- proxy
    if config.proxies and config.proxies[scheme] then
        host = config.proxies[scheme].host
        port = config.proxies[scheme].port
    end

    local ok, err = sock:connect(host, port)
    if not ok then
        return nil, err
    end

    if scheme == "https" then
        local ssl = config.ssl
        local reused_session = false
        local server_name = host
        local verify = false

        if ssl then
            reused_session = ssl.reused_session
            server_name = ssl.server_name
            verify = ssl.verify
        end

        ctx.state = STATE.HANDSHAKE

        local ok, err = sock:sslhandshake(reused_session, server_name, verify)
        if not ok then
            return nil, err
        end
    end

    ctx.sock = sock

    return true
end


local function send(ctx)
    local url_parts = ctx.url_parts
    local config = ctx.config
    local path = url_parts.path
    local uri = path
    local args = url_parts.args
    local opts = ctx.opts

    if #url_parts.args > 0 then
        uri = str_format("%s?%s", uri, args)
    end

    -- send timeout
    ctx.sock:settimeout(config.timeouts[2])

    local header_lines = serialize(config.headers)
    local content = {
        ctx.method, " ", uri, " ", config.version,
        CRLF,
        header_lines,
    }

    ctx.state = STATE.SEND_HEADER
    local _, err = ctx.sock:send(content)
    if err then
        return nil, err
    end

    local body = config.body
    if not body then
        return true
    end

    ctx.state = STATE.SEND_BODY

    if not util.is_func(body) then
        local bytes, err = ctx.sock:send(body)
        if not bytes then
            return nil, err
        end
    else
        repeat
            local chunk, err = body()
            if not chunk then
                return nil, err
            end

            local data

            if chunk == "" then
                data = "0\r\n\r\n"
            else
                data = str_format("%x\r\n%s\r\n", #chunk, chunk)
            end

            local bytes, err = ctx.sock:send(data)
            if err then
                return nil, err
            end

        until not chunk or chunk == ""
    end

    return true
end


local function close(r)
    r.ctx.state = STATE.CLOSE
    close_sock(r)
end


local function receive(ctx)
    local config = ctx.config
    local r = util.new_tab(0, 11)

    -- state
    ctx.state = STATE.RECV_HEADER

    -- read timeout
    ctx.sock:settimeout(config.timeouts[3])

    local line_reader = ctx.sock:receiveuntil(CRLF)
    ctx.line_reader = line_reader

    local status_line, err = line_reader()
    if not status_line then
        return nil, err
    end

    -- 0) ctx
    r.ctx = ctx

    -- 1) url
    r.url = ctx.url

    -- 2) method
    r.method = ctx.method

    -- 3) status line
    r.status_line = status_line
    local part, err = parse_status_line(status_line)
    if not part then
        return nil, err or "bad status line"
    end

    -- 4) status code
    r.status_code = part.status_code

    -- 5) http version
    r.http_version = part.http_version

    -- 6) headers
    r.headers = util.header_dict(nil, 0, 9)

    -- 7) close
    r.close = close

    while true do
        local line, err = line_reader()
        if not line then
            return nil, err
        end

        if line == "" then
            break
        end

        local name, value, err = parse_header_line(line)
        if err then
            return nil, err
        end

        if name and value then
            -- A recipient MAY combine multiple header fields with the same
            -- field name into one "field-name: field-value" pair,
            -- without changing the semantics of the message,
            -- by appending each subsequent field value
            -- to the combined field value in order, separated by a comma.

            name = str_lower(name)

            local ovalue = r.headers[name]
            if not ovalue then
                r.headers[name] = value

            elseif util.is_tab(ovalue) then
                tab_insert(r.headers[name], value)

            else
                r.headers[name] = util.new_tab(4, 0)
                r.headers[name][1] = ovalue
                r.headers[name][2] = value
            end
        end
    end

    local ok, err = adjust_response_headers(r)
    if not ok then
        return nil, err
    end

    adjust_body(r)

    return r
end


local function send_request(ctx)
    local url_parts = ctx.url_parts
    local error_filter = ctx.config.error_filter

    ctx.state = STATE.READY

    -- 1) connect
    local ok, err = connect(ctx)
    if not ok then
        if error_filter then
            error_filter(ctx.state, err)
        end

        return nil, err
    end

    -- 2) send
    local ok, err = send(ctx)
    if not ok then
        if error_filter then
            error_filter(ctx.state, err)
        end

        return nil, err
    end

    -- 3) receive
    local r, err = receive(ctx)
    if not r then
        if error_filter then
            error_filter(ctx.state, err)
        end

        return nil, err
    end

    return r
end


local function redirect(r)
    r:close()

    local ctx = r.ctx
    local config = ctx.config

    local url = adjust_location(r)
    if not url then
        return nil, "no location header"
    end

    -- fix the Host header when the origin is not same.
    config.headers["Host"] = ctx.url_parts.host

    -- FIXME correct the method if necessary
    local next_r, err = _M.request(ctx.method, url, config)
    if not next_r then
        return nil, err
    end

    return next_r
end


_M.request = function(method, url, opts)
    local url_parts, err = parse_url(url)
    if not url_parts then
        return nil, err or "bad url"
    end

    local config = util.config(opts)

    local chunk = {
        size = -1,
        rest = 0,
        leave = false,
    }

    -- request context
    local ctx = {
        method      = method,
        url         = url,
        url_parts   = url_parts,
        config      = config,
        state       = STATE.UNREADY,
        sock        = nil,
        chunk       = chunk,
        line_reader = nil,
    }

    adjust_request_headers(ctx)

    local r, err = send_request(ctx)
    if not r then
        return nil, err
    end

    local hooks = config.hooks
    if hooks and hooks.response then
        hooks.response(r)
    end

    if allow_redirects(r) then
        return redirect(r)
    end

    return r
end


_M.register = function(obj, session)
    for _, method in ipairs(HTTP_METHODS) do
        local name = str_lower(method)
        obj[name] = function(url, opts)
            return _M.request(method, url, opts)
        end
    end
end


_M.state = function(state)
    return STATE_NAME[state] or "unknown"
end


_M.STATE = STATE

return _M
